File: collector/destinations/stdout.go
// Package destinations contains the built-in destination types. Adding a
// new destination type means adding one file to this package: implement
// core.Destination and register a constructor in an init function.
package destinations

import (
	"context"
	"encoding/json"
	"os"

	"github.com/pcarpe4/bento/collector/core"
)

// stdoutDestination writes each message as a JSON line to standard output.
type stdoutDestination struct {
	encoder *json.Encoder
}

func init() {
	core.RegisterDestination("stdout", func(cfg core.Fields) (core.Destination, error) {
		return &stdoutDestination{encoder: json.NewEncoder(os.Stdout)}, nil
	})
}

func (s *stdoutDestination) Write(ctx context.Context, batch []core.Message) error {
	for _, msg := range batch {
		if err := s.encoder.Encode(msg); err != nil {
			return err
		}
	}
	return nil
}

func (s *stdoutDestination) Close(ctx context.Context) error {
	return nil
}

File: collector-rs/src/destinations/stdout.rs
use anyhow::Result;
use async_trait::async_trait;
use tokio::io::AsyncWriteExt;

use crate::message::Message;
use crate::registry::Destination;

/// Writes each message as a JSON line to standard output.
struct StdoutDestination;

pub fn new(_raw: &serde_yaml::Value) -> Result<Box<dyn Destination>> {
    Ok(Box::new(StdoutDestination))
}

#[async_trait]
impl Destination for StdoutDestination {
    async fn write(&self, batch: &[Message]) -> Result<()> {
        let mut out = tokio::io::stdout();
        for msg in batch {
            let mut line = serde_json::to_vec(msg)?;
            line.push(b'\n');
            out.write_all(&line).await?;
        }
        out.flush().await?;
        Ok(())
    }
}
